File: 151-reverse-words-in-a-string/reverse-words-in-a-string.py
class Solution:
    def reverseWords(self, s: str) -> str:
        words = []
        i = 0
        n = len(s)
        while i < n:
            while i < n and s[i] == ' ':
                i += 1
            if i >= n:
                break;
            start = i
            while i < n and s[i] != ' ':
                i += 1
            words.append(s[start:i])
        return ' '.join(reversed(words))

File: 151-reverse-words-in-a-string/reverse-words-in-a-string.cpp
class Solution {
public:
    string reverseWords(string s) {
        vector<string> words;
        int n = s.size();
        int i = 0;

        while (i < n) {
            while (i < n && s[i] == ' ')
                i++;
            if (i >= n)
                break;
            int j = i;
            while (j < n && s[j] != ' ')
                j++;
            words.push_back(s.substr(i, j - i));
            i = j;
        }

        string result;
        for (int k = words.size() - 1; k >= 0; --k) {
            result += words[k];
            if (k > 0)
                result += " ";
        }
        return result;
    }
};